### ISSUE TITLE
Fast-Tokenizer: loss of accuracy

### ISSUE THREAD

**Issue by BrunoFANG1** [Physical-Intelligence/openpi#485]
I just played around with FAST-Tokenizer, and tested it with different action inputs. I compared the decoded action and original input action and found significant errors. Can someone give me a hint on this? I think for some tasks that require precision, it matters a lot.

BTW, I have tried min-max normalization, quantile normalization, and mean-var normalization, all of them show errors over 2cm along position x, position-y, and position-z.

Is there an explanation to that or I just need to train it from scratch.

**Comment by kpertsch** [comment]
Hi! FAST is a lossy tokenizer, since we use compression -- but the level of compression is a hyperparameter you can choose.
Concretely, the lossy bit in FAST is the quantization of the DCT coefficient matrix. We implement this through a scale-and-round operation. Here, the scale parameter is what trades off between compression and accuracy -- larger scale will give higher accuracy, but longer token sequences (ie less compression).
I would suggest that you try training a policy with the default parameters first -- we have used these for all the tasks we are running at pi, so I think they should still work for fine-grained manipulation. If however, you find that they don't work well (eg indicated by a clear performance difference between the diffusion pi0 and pi0-FAST finetunes), then you can try to increase the scale parameter and see whether that resolves the problem for you.

Hope this helps! Will close the issue for now!
